File: backend/scala-seed-project/src/middleware/authorizeMiddleware.scala
// src/middleware/authorizeMiddleware.scala

import cats.effect.Sync
import org.http4s.{Request, Response}
import org.http4s.dsl.Http4sDsl
import org.http4s.HttpRoutes
import scala.concurrent.ExecutionContext.Implicits.global

object AuthorizeMiddleware {
  def authorizeMiddleware[F[_]: Sync](requiredPermissions: List[String])(routes: HttpRoutes[F]): HttpRoutes[F] = {
    val dsl = new Http4sDsl[F]{}
    import dsl._

    HttpRoutes.of[F] {
      case req =>
        // Retrieve the role from the session
        val userRole = req.attributes.get(UserSessionAttr).map(_.role) // Assume UserSessionAttr holds the role

        userRole match {
          case Some(role) if role == "admin" =>
            routes(req) // Allow access for admin role
          
          case Some(role) =>
            // Logic to fetch permissions for the user's role (assume getRolePermissions is defined)
            val rolePermissions = getRolePermissions(role) // Function to get permissions from database

            // Check if the role has the required permission(s)
            val hasPermission = requiredPermissions.exists(rolePermissions.contains)

            if (hasPermission) {
              routes(req) // Proceed to the next route if permission is granted
            } else {
              Forbidden("Forbidden: You do not have access to this resource")
            }

          case None => Forbidden("Role not found in session")
        }
    }
  }

  // Placeholder for the function that gets role permissions
  def getRolePermissions(role: String): List[String] = {
    // Logic to fetch role permissions from the database
    // This is a placeholder; replace with actual DB call
    List("view", "edit") // Example permissions
  }
}

File: backend/node_version/src/middleware/authorizeMiddleware.ts
// backend\src\middleware\authorizeMiddleware.ts
import { Request, Response, NextFunction } from 'express';
import Role from '../models/roleModel';

// Middleware to authorize based on permissions
export const authorizeMiddleware = (requiredPermissions: string[]) => {
  return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    // Retrieve the role from the session
    const userRole = req.session.user?.role; // Ensure that user role is stored in session
    try {
      // Check if user role exists in the session
      if (!userRole) {
        res.status(403).json({ message: 'Role not found in session' });
        return; // Explicitly return after sending response
      }

      // Find the role based on the userRole from the session
      const role = await Role.findOne({ where: { name: userRole } });

      if (!role) {
        res.status(403).json({ message: 'Role not found' });
        return; // Explicitly return after sending response
      }

      const rolePermissions = role.permissions;
      // Admins can do anything, so short-circuit for the "admin" role
      if (userRole == 'admin' || rolePermissions.includes('all')) {
        return next(); // Allow access
      }

      // Check if the role has the required permission(s)
      const hasPermission = requiredPermissions.some(permission => rolePermissions.includes(permission));
      if (!hasPermission) {
        res.status(403).json({ message: 'Forbidden: You do not have access to this resource' });
        return; // Explicitly return after sending response
      }

      // Proceed if the user has the required permissions
      next();
    } catch (error: any) {
      res.status(500).json({ message: 'Authorization failed', error: error.message });
    }
  };
};
